### PR TITLE
feat: Surface Pod warning events onto TaskRun status

### LINKED ISSUE #10373: Surface Pod infrastructure failure reasons (from Events) onto TaskRun status
### Feature request

Surface actionable infrastructure failure reasons (from Pod Events) onto the
`TaskRun` status when a Pod is stuck starting and the Pod object itself carries
no useful information.

When a Pod fails to *start* (as opposed to a Step failing at runtime), there are
no container logs, because no container ever runs. In many of these cases the
Pod object contains enough for Tekton to explain the problem, and Tekton already
does a good job here in `pkg/pod/status.go` (`updateIncompleteTaskRunStatus`):

- Unschedulable due to insufficient resources → `ExceededNodeResources`
- Missing Secret/ConfigMap referenced via `envFrom` / `env.valueFrom`, or bad
  container config → `CreateContainerConfigError` (with a message)
- `ImagePullBackOff` / `ErrImagePull` / `InvalidImageName` / … → `PullImageFailed`
- Otherwise `getWaitingMessage()` falls back to the container waiting message,
  then to non-`True` Pod conditions, then to `pod.Status.Message`.

There is, however, a category of failures where **the actionable detail exists
only as a Kubernetes Event and is never written to any field of the Pod
object**. The canonical example is a Secret/ConfigMap that is missing but
mounted as a **volume** (rather than referenced via env). In that case:

- `pod.status.phase` stays `Pending`
- the container is stuck with `state.waiting.reason: ContainerCreating` and an
  **empty** message
- Pod conditions only say `containers with unready status: [step-x]`
- the real reason is emitted only as a `Warning` Event with reason `FailedMount`

As a result, `getWaitingMessage()` can only produce something like
`build step "step-x" is pending with reason "ContainerCreating"`, and the
TaskRun sits in that state until it times out with no indication of the cause.

The missing-Secret-volume case is the cleanest example, but the same problem
(detail lives only in an Event; Pod shows generic `Pending`/`ContainerCreating`)
applies to several other failure modes:

- **Volume mount/attach** (all volume types): `FailedMount` /
  `FailedAttachVolume` ("Volume is already exclusively attached to one node",
  mount timeouts).
- **PVC not bound / not provisioned**: `FailedScheduling` ("pod has unbound
  immediate PersistentVolumeClaims") or PVC `ProvisioningFailed`.
- **CNI / sandbox**: `FailedCreatePodSandBox` ("failed to set up sandbox
  container … network …") — no container status at all.
- **Scheduling detail**: `FailedScheduling` ("0/N nodes are available: … had
  untolerated taint / didn't match node affinity …"); only the terse
  `Unschedulable` condition is on the Pod (insufficient-resources is already
  handled by Tekton).

A related but distinct case is **admission/quota rejection at pod-creation time**
(`ExceededResourceQuota`, PodSecurity/LimitRange/webhook denials): here the Pod
may never be created, so there is no Pod UID to look up — the event has to be
read from the pod-create call / owning object. Worth handling as a separate
sub-problem.

#### Reproduction

A Pod that mounts a non-existent Secret as a volume:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: missing-secret-volume
spec:
  containers:
  - name: step
    image: registry.k8s.io/pause:3.9
    volumeMounts:
    - name: vol
      mountPath: /etc/foo
  volumes:
  - name: vol
    secret:
      secretName: does-not-exist
```

The Pod status (trimmed) — note the empty waiting message and generic conditions:

```yaml
status:
  phase: Pending
  conditions:
  - type: Ready
    status: "False"
    reason: ContainersNotReady
    message: 'containers with unready status: [step]'
  - type: ContainersReady
    status: "False"
    reason: ContainersNotReady
    message: 'containers with unready status: [step]'
  - type: PodScheduled
    status: "True"
  containerStatuses:
  - name: step
    state:
      waiting:
        reason: ContainerCreating   # <- no message
```

The only place the cause appears:

```
$ kubectl get events --field-selector involvedObject.uid=<pod-uid>,type=Warning
Warning  FailedMount  MountVolume.SetUp failed for volume "vol" : secret "does-not-exist" not found
```

### Use case

A user submits a PipelineRun that references a Secret/ConfigMap as a workspace
or volume that doesn't exist (typo, not yet created, wrong namespace). The
TaskRun's Pod never starts. Today the TaskRun stays `Running`/`Pending` with a
message like `build step "step-x" is pending with reason "ContainerCreating"`
until it times out — with no logs and no reason. The user (or their support
team) has to reach for `kubectl describe pod` / `kubectl get events`, which:

- requires cluster access the user may not have (shared/multi-tenant clusters),
- requires Kubernetes expertise to interpret,
- is often impossible after the fact, because Events may be short-lived.

If Tekton copied the distilled reason onto the TaskRun status, the information
would be visible through the normal status the user already consumes, and — 
importantly — it would be **persisted** with the TaskRun (and anything that
archives TaskRuns, e.g. Results), so it survives long after the Event is gone.

### Proposed (low-cost) implementation

Only in the case where the Pod is `Pending`/`ContainerCreating` **and** the Pod
object yields no useful message, do a single, targeted Events **List** scoped to
that one Pod, and surface the most recent `Warning` event's reason + message
onto the TaskRun status condition.

Extend the `default` branch of `updateIncompleteTaskRunStatus` in
`pkg/pod/status.go`:

```go
default:
    msg := getWaitingMessage(pod)
    if isGenericPending(msg) { // e.g. reason "ContainerCreating" with no detail
        if evReason, evMsg := latestWarningEvent(ctx, kubeclient, pod); evMsg != "" {
            msg = fmt.Sprintf("%s: %s", evReason, evMsg)
        }
    }
    markStatusRunning(trs, ReasonPodPending, msg)
```

where the lookup is a field-selected List for the specific Pod (no informer, no
watch):

```go
events, _ := kube.CoreV1().Events(pod.Namespace).List(ctx, metav1.ListOptions{
    FieldSelector: fmt.Sprintf("involvedObject.uid=%s,type=Warning", pod.UID),
})
```

Why this is intentionally cheap:

- **Opt-in first.** Ship it behind a feature flag (e.g. `surface-pod-events`,
  disabled by default) so clusters can enable it deliberately, and we can
  validate behaviour/cost before considering it on-by-default.
- **No new watch/informer on Events.** The existing Pod informer already
  triggers a TaskRun reconcile when the Pod changes; we merely add a
  per-Pod List, and only when the Pod is stuck with no useful status. This
  avoids the well-known etcd pressure of watching Events cluster/namespace wide.
- **Nothing new is exposed to end users** beyond a single curated line placed on
  the TaskRun status. The controller decides what to copy, so it can be limited
  to `Warning` events / an allowlist of reasons if desired.
- **Persistence for free:** because it lands on `TaskRun.status`, the reason
  outlives the Event, whose retention is cluster-dependent and may be short.

Implementation notes / open questions:

- RBAC: the controller ClusterRole currently has only `create/update/patch` on
  `events`; this needs `get`/`list` added.
- Consider gating behind a feature flag (e.g. `surface-pod-events`) for a safe,
  opt-in rollout.
- Decide the selection policy when multiple Warning events exist (most recent,
  or a reason allowlist such as `FailedMount`, `FailedScheduling`,
  `FailedCreatePodSandBox`).
- Bound the message length written to status.

### Alternatives to consider

1. **Enrich only from the Pod object (no events at all).** Free and simplest,
   but as shown above it cannot cover the volume-mount case, because the detail
   is never written to the Pod — so this is insufficient on its own. Still worth
   doing for the cases where the Pod *does* carry a message.
2. **Watch/informer on Events.** More responsive and could reflect updates
   continuously, but introduces ongoing etcd load and complexity — the opposite
   of the goal.
3. **A dedicated `infra`/`events` sub-resource or field on TaskRun status**
   (structured list of relevant events) instead of folding into the condition
   message. Richer for UIs, but larger API surface and status size; could be a
   follow-up if the single-message approach proves too limited.
4. **Leave it to external tooling** (dashboards/log collectors) to correlate
   events. Doesn't solve the "no access / no persistence / needs k8s expertise"
   problems for the user looking at the TaskRun.

### PR DESCRIPTION
**Pull request by KMI1011**
# Changes

Fixes #10373

When a Pod is stuck `Pending` with `ContainerCreating` and no useful message
(e.g. a Secret mounted as a volume doesn't exist), the actionable reason
exists only in Kubernetes Events — never on the Pod object. Today the TaskRun
sits with a generic `"Pending"` message until timeout.

This PR adds a new alpha feature flag `surface-pod-events` (disabled by
default) that, when enabled, performs a single field-selected Events List
scoped to the stuck Pod's UID. If a Warning event is found (e.g.
`FailedMount`, `FailedScheduling`, `FailedCreatePodSandBox`), its reason and
message are surfaced onto the TaskRun status condition.

**What changed:**

- **Feature flag**: Added `surface-pod-events` alpha flag to
  `pkg/apis/config/feature_flags.go` and `config/config-feature-flags.yaml`.
- **Event lookup**: Introduced `pkg/pod/events.go` with `latestWarningEvent`,
  `isGenericPending`, and `formatEventMessage` helpers.
- **Status wiring**: Extended `updateIncompleteTaskRunStatus` in
  `pkg/pod/status.go` to look up Pod events when the flag is enabled and the
  Pod is in a generic pending state.
- **RBAC**: Added `get`, `list` to the controller's `events` permissions in
  `config/200-clusterrole.yaml`.
- **Tests**: Added unit tests for all new helpers (`pkg/pod/events_test.go`),
  integration tests for the feature-flagged status path
  (`pkg/pod/status_test.go`), and feature flag parsing tests.

**Design decisions:**
- No informer or watch on Events — single targeted List per reconcile, only
  when the Pod is stuck with no useful status.
- Event message bounded to 1024 characters to avoid etcd bloat.
- Most-recent Warning event is selected when multiple exist.
- Errors from the Events API are logged at debug level and silently fall back
  to the existing generic message.

# Submitter Checklist

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Surface Pod infrastructure failure reasons (from Warning events such as
FailedMount, FailedScheduling, FailedCreatePodSandBox) onto the TaskRun
status condition when a Pod is stuck pending with no useful message. Gated
behind the new `surface-pod-events` alpha feature flag (disabled by default).
```

### CODE REVIEW

File: pkg/pod/events.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	for _, s := range pod.Status.InitContainerStatuses {
+		if s.State.Waiting != nil && s.State.Waiting.Reason == "ContainerCreating" && s.State.Waiting.Message == "" {
+			return true
+		}
+	}
+	// Also treat the bare "Pending" fallback from getWaitingMessage as generic:
+	// this happens when no container has a waiting message AND all pod
+	// conditions are True AND pod.Status.Message is empty.
+	// Guard against pods with no container statuses at all (e.g. not yet scheduled).
```
Comment (aThorp96) [inline pkg/pod/events.go]: Not-scheduled I think is a valid "GenericPending" case. If the pod can't be scheduled on a node it will get stuck into the failure mode we're hoping to detect here.

File: pkg/pod/status.go [hunk 2]
```diff
@@ -631,14 +631,19 @@ func updateIncompleteTaskRunStatus(trs *v1.TaskRunStatus, pod *corev1.Pod) {
...
 			markStatusFailure(trs, ReasonCreateContainerConfigError, "Failed to create pod due to config error")
 		case isPullImageError(pod):
 			markStatusRunning(trs, ReasonPullImageFailed, getWaitingMessage(pod))
 		default:
-			markStatusRunning(trs, ReasonPodPending, getWaitingMessage(pod))
+			msg := getWaitingMessage(pod)
+			if config.FromContextOrDefaults(ctx).FeatureFlags.EnableSurfacePodEvents && isGenericPending(pod) {
+				if evReason, evMsg := latestWarningEvent(ctx, logger, kubeclient, pod); evMsg != "" {
+					msg = formatEventMessage(evReason, evMsg)
+				}
```
Comment (aThorp96) [inline pkg/pod/status.go]: We can inline this string building and simplify a bit  I think. 
Also, I believe the condition is supposed to be `eventReason != ""` instead of `eventMsg != ""`
```suggestion
				if eventReason, eventMsg := latestWarningEvent(ctx, logger, kubeclient, pod); eventReason != "" {
					msg = eventReason
					if eventMsg = strings.TrimSpace(eventMsg); eventMsg != "" {
						msg += ": " + eventMsg
					}
				}
```

File: pkg/pod/events.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+		msg = msg[:maxEventMessageLength] + "..."
+	}
+	return latest.Reason, msg
+}
+
+// isGenericPending returns true when the Pod is stuck in a pending state that
+// carries no actionable detail — the container waiting reason is
+// "ContainerCreating" with an empty message, or getWaitingMessage fell through
+// to one of its generic fallbacks.
+func isGenericPending(pod *corev1.Pod) bool {
```
Comment (aThorp96) [inline pkg/pod/events.go:70]: We've already done some of this work in `getWaitingMessage()` before this is called. Can we get away with just checking if `getWaitingMessage(pod) != "Pending"` or are there other conditions we need to rule out?

File: pkg/pod/events.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const maxEventMessageLength = 1024
+
+// latestWarningEvent returns the reason and message of the most recent Warning
+// event for the given Pod. It performs a single, field-selected List scoped to
+// the Pod's UID. If no Warning events exist or the lookup fails, it returns
+// empty strings.
```
Comment (aThorp96) [inline pkg/pod/events.go]: Nit: this is an implementation detail and isn't necessary to add to the comment
```suggestion
// latestWarningEvent returns the reason and message of the most recent Warning
// event for the given Pod. If no Warning events exist or the lookup fails, it returns
// empty strings.
```

File: pkg/pod/events.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+func latestWarningEvent(ctx context.Context, logger *zap.SugaredLogger, kubeclient kubernetes.Interface, pod *corev1.Pod) (reason, message string) {
+	if pod.UID == "" {
+		return "", ""
+	}
+	events, err := kubeclient.CoreV1().Events(pod.Namespace).List(ctx, metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("involvedObject.uid=%s,type=Warning", pod.UID),
+	})
+	if err != nil {
+		logger.Debugw("Failed to list pod warning events", "pod", pod.Name, "namespace", pod.Namespace, "error", err)
+		return "", ""
```
Comment (aThorp96) [inline pkg/pod/events.go:48]: Optional suggestion. No need to pass the logger all the way down for one debug log. Getting the logger from the context should work here. 
```suggestion
	if err != nil {
		logging.FromContext(ctx).Debugw("Failed to list pod warning events", "pod", pod.Name, "namespace", pod.Namespace, "error", err)
		return "", ""
```

File: pkg/pod/events.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	latest := events.Items[0]
+	for i := range events.Items[1:] {
+		ev := &events.Items[i+1]
+		if ev.LastTimestamp.After(latest.LastTimestamp.Time) {
+			latest = *ev
+		}
+	}
```
Comment (aThorp96) [inline pkg/pod/events.go]: This works great. The `i+1` can be a bit footgun though. There's a couple safer or more pragmatic ways to do this I would recommend 

Using a similar approach, but initialize the `latest` in the loop so you don't need to get clever with slices:
```go
	var latest corev1.Event
	for _, event := range events.Items {
		if latest == nil || event.LastTimestamp.After(latest.LastTimestamp.Time) {
			latest = *event
		}
	}
```

Using the standard library function `slices.MaxFunc()`
```go
	latest := slices.MaxFunc(
		events.Items,
		func(a corev1.Event, b corev1.Event) int {
			return a.LastTimestamp.Compare(b.LastTimestamp.Time)
		},
	)
```

File: pkg/pod/events.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+	msg := latest.Message
+	if len(msg) > maxEventMessageLength {
+		msg = msg[:maxEventMessageLength] + "..."
```
Comment (aThorp96) [inline pkg/pod/events.go]: Two things to note here:
1. Go strings are Runes not bytes, but slicing a string is on byte boundaries. If there's unicode in the string at the slice point it would get munged. I don't think that's a big concern in this case though. We probably don't need to worry about it; just calling it out. 
2. Since the `"..."` takes up three characters (in this case one byte each), I think you want `msg = msg[:maxEventMessageLength-3] + "..."`

File: pkg/pod/events.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	// Also treat the bare "Pending" fallback from getWaitingMessage as generic:
+	// this happens when no container has a waiting message AND all pod
+	// conditions are True AND pod.Status.Message is empty.
+	// Guard against pods with no container statuses at all (e.g. not yet scheduled).
+	if len(pod.Status.ContainerStatuses) == 0 && len(pod.Status.InitContainerStatuses) == 0 {
+		return false
+	}
+	if noContainerHasWaitingMessage(pod) && allConditionsTrue(pod) && pod.Status.Message == "" {
+		return true
```
Comment (aThorp96) [inline pkg/pod/events.go:92]: Can you add a brief comment with what these conditions are ruling out?

**Review by aThorp96 — COMMENTED** [review]
Thanks @KMI1011 ! I haven't read through the tests yet but I wanted to go ahead and share some feedback on the main changes

File: pkg/pod/events.go (hunk 3 above)
Comment (KMI1011) [reply pkg/pod/events.go:70]: I dont think so (?) `getWaitingMessage` conflates two different situations into non-`"Pending" `strings that are still generic from an event-surfacing perspective. `isGenericPending` catches cases like `ContainerCreating` with an empty message, which `getWaitingMessage` would mask behind a non-`"Pending"` but still uninformative string. Plus it's a cheap struct check that avoids wasted formatting work.

File: pkg/pod/events.go (hunk 1 above)
Comment (aThorp96) [reply pkg/pod/events.go]: It looks like the "not yet scheduled" example was removed from the comment but the semantics are still the same. If the pod is not scheduled, I think that's the type of circumstance we're trying to surface in the TaskRun status. Is there a reason we return `false` in that instance? Can we switch to returning `true`?

File: pkg/pod/status.go (hunk 2 above)
Comment (aThorp96) [reply pkg/pod/status.go]: @KMI1011 any thoughts about inlining the message building here?

File: pkg/pod/events.go (hunk 6 above)
Comment (aThorp96) [reply pkg/pod/events.go]: @KMI1011 can you please switch this loop to one of the safer approaches mentioned above?

File: pkg/pod/events.go (hunk 7 above)
Comment (aThorp96) [reply pkg/pod/events.go]: @KMI1011 (1) is general guidance but I believe (2) is a subtle logical bug

**Comment by aThorp96** [comment]
@KMI1011 Please also format the code with `make fmt` and update the generated code with `make generated`

File: pkg/pod/status.go (hunk 2 above)
Comment (KMI1011) [reply pkg/pod/status.go]: Good call, I think you're right. `evMsg != ""` was silently dropping things like a bare `FailedScheduling`. I'll make sure to inline it

File: pkg/pod/events.go (hunk 1 above)
Comment (KMI1011) [reply pkg/pod/events.go]: Yup, I can do that!

**Comment by Omeramsc** [comment]
Hey @abayer @dibyom , PTAL

**Review by aThorp96 — COMMENTED** [review]
/lgtm

**Comment by aThorp96** [comment]
@KMI1011 Please address the linting errors: https://github.com/tektoncd/pipeline/actions/runs/32054582059/job/96815958767?pr=10546#step:5:245

File: pkg/pod/status.go [hunk 9]
```diff
@@ -631,14 +631,22 @@ func updateIncompleteTaskRunStatus(trs *v1.TaskRunStatus, pod *corev1.Pod) {
...
-			// if subPath directory creation errors, mark as running and wait for recovery
 			markStatusRunning(trs, ReasonPodPending, "Waiting for subPath directory creation to complete")
 		case isPodHitConfigError(pod):
 			markStatusFailure(trs, ReasonCreateContainerConfigError, "Failed to create pod due to config error")
 		case isPullImageError(pod):
 			markStatusRunning(trs, ReasonPullImageFailed, getWaitingMessage(pod))
 		default:
-			markStatusRunning(trs, ReasonPodPending, getWaitingMessage(pod))
+			msg := getWaitingMessage(pod)
+			if config.FromContextOrDefaults(ctx).FeatureFlags.EnableSurfacePodEvents && isGenericPending(pod) {
```
Comment (Omeramsc) [inline pkg/pod/status.go:641]: What happens once the surfaced event ages out of etcd but the Pod is still stuck? On the next reconcile, won't `latestWarningEvent` return nothing (and the status message would revert)  losing the reason that was already correctly surfaced?

File: pkg/pod/events.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+	}
+	for _, s := range pod.Status.InitContainerStatuses {
+		if s.State.Waiting != nil && s.State.Waiting.Reason == "ContainerCreating" && s.State.Waiting.Message == "" {
+			return true
+		}
+	}
+	// No container statuses means the kubelet has not started on this Pod, so it
+	// has not been scheduled onto a node. Whatever is blocking scheduling is
+	// reported only via Warning events such as FailedScheduling.
+	if len(pod.Status.ContainerStatuses) == 0 && len(pod.Status.InitContainerStatuses) == 0 {
```
Comment (Omeramsc) [inline pkg/pod/events.go:84]: This is true for every Pod in the first reconcile after creation, before the kubelet has reported any container status, including pods that are about to schedule and start completely normally. Are we firing the Events list of every TaskRun pod's startup then?

File: pkg/pod/status.go (hunk 9 above)
Comment (aThorp96) [reply pkg/pod/status.go:641]: I don't know how often this will happen in practice, since IME these types of events are re-emitted every schedule attempt. I think it's reasonable to set this up so when the pod would get a "generic pending message" but the existing status indicates there was "generic pending with warning" message, then the previous message isn't overwritten. The issue that _might_ get into is ephemeral issues could become sticky. I don't know which is the worse tradeoff though. @vdemeester any thoughts?

**Comment by aThorp96** [comment]
Discussed with @KMI1011 a bit ago and as they now have less bandwidth to continue work on this PR @Omeramsc or I will be carrying it forward

File: pkg/pod/status.go (hunk 9 above)
Comment (waveywaves) [inline pkg/pod/status.go:641]: If the Warning arrives after this reconcile, what triggers another lookup? The controller doesn't watch Events, and the next scheduled reconcile may be the timeout, which runs before this path. Could we add a bounded retry and test Event arrival after the first empty lookup?

File: pkg/pod/events.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+	if err != nil {
+		logging.FromContext(ctx).Debugw("Failed to list pod warning events", "pod", pod.Name, "namespace", pod.Namespace, "error", err)
+		return "", ""
+	}
+	if len(events.Items) == 0 {
+		return "", ""
+	}
+
+	latest := slices.MaxFunc(events.Items, func(a, b corev1.Event) int {
+		return a.LastTimestamp.Compare(b.LastTimestamp.Time)
```
Comment (waveywaves) [inline pkg/pod/events.go:55]: `LastTimestamp` is optional/legacy; modern Events use `EventTime` and `Series.LastObservedTime`. When it is unset, selection depends on list order. Could we use the available timestamp fields and add modern Event tests?

File: pkg/pod/events.go [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// isGenericPending returns true when the Pod is stuck in a pending state that
+// carries no actionable detail — the container waiting reason is
+// "ContainerCreating" with an empty message, the Pod has not been scheduled
+// onto a node yet, or getWaitingMessage fell through to one of its generic
+// fallbacks.
+func isGenericPending(pod *corev1.Pod) bool {
+	for _, s := range pod.Status.ContainerStatuses {
+		if s.State.Waiting != nil && s.State.Waiting.Reason == containerWaitingReasonCreating && s.State.Waiting.Message == "" {
```
Comment (waveywaves) [inline pkg/pod/events.go:72]: This can return `true` despite useful container, Pod-condition, or Pod-status detail, which the caller then replaces with a possibly stale Event. Could we query Events only when all Pod-derived detail is generic and add a mixed-status test?

File: config/200-clusterrole.yaml [hunk 13]
```diff
@@ -62,10 +62,11 @@ rules:
   - apiGroups: [""]
     resources: ["pods", "persistentvolumeclaims"]
     verbs: ["get", "list", "create", "update", "delete", "patch", "watch"]
-  # Write permissions to publish events.
+  # Read/write permissions for events: create/update/patch to publish events,
+  # get/list to surface Pod warning events onto TaskRun status (surface-pod-events flag).
   - apiGroups: [""]
     resources: ["events"]
-    verbs: ["create", "update", "patch"]
+    verbs: ["get", "list", "create", "update", "patch"]
```
Comment (waveywaves) [inline config/200-clusterrole.yaml:69]: The implementation only lists Events. Can we omit the unused `get` permission?

File: pkg/pod/status.go (hunk 9 above)
Comment (waveywaves) [reply pkg/pod/status.go:641]: This diagnosis isn't durable: an expired/missing Event resets it to `Pending`, and timeout drops previous `Pending` messages, so the final TaskRun/Results record loses it. Could we preserve Event-derived context and test both expiry and timeout?

File: pkg/pod/events.go (hunk 10 above)
Comment (waveywaves) [reply pkg/pod/events.go:84]: A new Pod normally has no container statuses, so every startup takes this path and may issue repeated uncached Event LISTs. Could we narrow or debounce the lookup and assert a bounded request count?

**Comment by Omeramsc** [comment]
@waveywaves As mentioned by @aThorp96 , I continued the work on this feature on #10690 , PTAL there. This PR can be safely closed.

File: config/200-clusterrole.yaml (hunk 13 above)
Comment (Omeramsc) [reply config/200-clusterrole.yaml:69]: removed on 10373

File: pkg/pod/events.go (hunk 11 above)
Comment (Omeramsc) [reply pkg/pod/events.go:55]: Fixed on 10690, events are ordered using: `series.lastObservedTime` -> `eventTime` -> `lastTimestamp` -> Event name

File: pkg/pod/events.go (hunk 12 above)
Comment (Omeramsc) [reply pkg/pod/events.go:72]: Replaced with a function that examines multiple factors instead of relying on the first find (regular containers, waiting states, PodStatus.Message, etc). Event lookup will happen only when all available Pod information is generic.

File: pkg/pod/status.go (hunk 9 above)
Comment (Omeramsc) [reply pkg/pod/status.go:641]: The surfaced event text is now labeled with a stable prefix (`Last observedf pod warning: .....`). On a later generic reconcile, it recognize this message and preserve it instead of reverting to `Pending`.

Comment (Omeramsc) [reply pkg/pod/status.go:641]: Added a scheduled two bounded retries. if the Event still is absent, the controller performs a final retry after 30 seconds, then stops looking for that Pod's UID.

File: pkg/pod/events.go (hunk 10 above)
Comment (Omeramsc) [reply pkg/pod/events.go:84]: added a retry state on `taskrun.go` to resolve this; now it is keyed by the Pod's UID, and have 3 retries.

**Comment by aThorp96** [comment]
Thanks @Omeramsc and @KMI1011. Closing in favor of https://github.com/tektoncd/pipeline/pull/10690
